Classes/Controller/FormController.php

```diff
@@ -615,6 +615,10 @@ private function formStepChange(): void {
     }
 
     $this->formConfig->session->set('step', $this->formConfig->step);
+
+    if ('json' == $this->formConfig->responseType) {
+      $this->jsonResponse->step = $this->formConfig->step;
```

Comment:
I'm no fan of doing this...

We should add a new function 

```php
  private function prepareJsonResponse(): string {
    $this->jsonResponse->fieldsErrors = $this->formConfig->fieldsErrors;
    $this->jsonResponse->fieldSets = $this->formConfig->fieldSets;
    $this->jsonResponse->formErrors = $this->formConfig->formErrors;
    $this->jsonResponse->formValues = $this->formConfig->formValues;
    $this->jsonResponse->step = $this->formConfig->step;

    return json_encode($this->jsonResponse) ?: '{}';
  }
```

Remove this lines here and change `public function formAction()`.


Line 435:

```php
return $this->jsonResponse($this->prepareJsonResponse());
```

Line 443

```php
return $this->jsonResponse($this->prepareJsonResponse());
```

Line 449

```php
      if ('json' == $this->formConfig->responseType) {
        return $this->jsonResponse($this->prepareJsonResponse());
      }
```

Line 465

```php
   if ('json' == $this->formConfig->responseType) {
      $this->jsonResponse->fileUpload = $this->formConfig->fileUpload;
      $this->jsonResponse->formUploads = $this->formConfig->formUploads;
      $this->formConfig->processDebugLog();

      return $this->jsonResponse($this->prepareJsonResponse());
    }
```

Reply:
That would probably be a better way of fixing that. I also just now noticed that there is more data missing when changing steps.
```
      $this->jsonResponse->fileUpload = $this->formConfig->fileUpload;
      $this->jsonResponse->formUploads = $this->formConfig->formUploads;
```
Is not run when changing a step, thus fileuploads are not accessible in the frontend.

`this->prepareFormSets();` also wont get called which means that metadata fields cant get updated. I'll also add that i guess